### PR TITLE
chore: support scaling of the resources in the ui

### PR DESCRIPTION
**Pull request by irinam0992**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates? **No**
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

# Scale up and down the resources in the UI

Screenshot 2023-09-22 at 14 15 44

## Summary

Implement UI where users could scale up and down the resources by clicking arrows or entering inputs.

## Motivation

The main motivation is to provide a convenient way to scale resources from ArgoCD UI, this can be useful for users who are not aware of yaml language and provide a user-friendly way to scale resources.

Enhancement proposal can be found by the following link:
https://github.com/argoproj/argo-cd/pull/15633

### CODE REVIEW

**Comment by mubarak-j** [comment]
Is it possible to add the lua action to Argo Rollout resource as well?

File: ui/src/app/applications/components/utils.tsx [hunk 1]
```diff
@@ -123,6 +124,74 @@ export async function deleteApplication(appName: string, appNamespace: string, a
     return confirmed;
 }
 
+export async function scalePods(resource: ResourceTreeNode, metadata: models.ObjectMeta, apis: ContextApis): Promise<boolean> {
```
Comment (crenshaw-dev) [inline ui/src/app/applications/components/utils.tsx:127]: imo, if at all possible, the UI should generically present "parameters" communicated by the action definition itself. Hard-coding a parameters UI for just the scale action significantly limits the power of this feature.

Comment (irinam0992) [reply ui/src/app/applications/components/utils.tsx:127]: I believe I understand your point.
But we cannot use the current logic with discovery actions.
Currently for all actions which returned from backend by getResourceActions function is built default ui which triggers this action and show confirmation popup.
In our case we need custom UI with pop up which you can see in the PR body and this solution does not work.

Comment (crenshaw-dev) [reply ui/src/app/applications/components/utils.tsx:127]: > we cannot use the current logic with discovery actions

The actions currently communicate up info like "disabled" or "not disabled."

I think we could augment that information with "has parameters," "list of parameters," "validation info for parameters," etc. The UI could react to that configuration and display the appropriate popup and inputs.

Comment (irinam0992) [reply ui/src/app/applications/components/utils.tsx:127]: @crenshaw-dev We can, but for scale case it is special and custom popup.
Screenshot 2023-09-19 at 12 01 17
How would you suggest to process this case

Comment (crenshaw-dev) [reply ui/src/app/applications/components/utils.tsx:127]: Let's consolidate the discussion over here: https://github.com/argoproj/argo-cd/pull/15633#discussion_r1334669882

Comment (irinam0992) [reply ui/src/app/applications/components/utils.tsx]: @crenshaw-dev 
Hi
Did you have to take a look at my changes?

File: pkg/apis/application/v1alpha1/types.go [hunk 2]
```diff
@@ -1984,11 +1984,13 @@ type ResourceActionDefinition struct {
...
-	Disabled    bool                  `json:"disabled,omitempty" protobuf:"varint,3,opt,name=disabled"`
-	IconClass   string                `json:"iconClass,omitempty" protobuf:"bytes,4,opt,name=iconClass"`
-	DisplayName string                `json:"displayName,omitempty" protobuf:"bytes,5,opt,name=displayName"`
+	Name          string                `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	Params        []ResourceActionParam `json:"params,omitempty" protobuf:"bytes,2,rep,name=params"`
+	Disabled      bool                  `json:"disabled,omitempty" protobuf:"varint,3,opt,name=disabled"`
+	IconClass     string                `json:"iconClass,omitempty" protobuf:"bytes,4,opt,name=iconClass"`
+	DisplayName   string                `json:"displayName,omitempty" protobuf:"bytes,5,opt,name=displayName"`
+	DefaultValue  string                `json:"defaultValue,omitempty" protobuf:"bytes,6,opt,name=defaultValue"`
+	HasParameters bool                  `json:"hasParameters,omitempty" protobuf:"bytes,7,opt,name=hasParameters"`
```
Comment (crenshaw-dev) [inline pkg/apis/application/v1alpha1/types.go:1993]: Don't wanna derail what you're doing, 'cause it's looking great. Just a question, could we possible infer this from the length of the Params field, or do we definitely need the new boolean?

**Comment by komapa** [comment]
Hey folks, this looks great. If this gets merged as fast as you expect it to be merged, which ArgoCD version do you think it will be available for? Appreciate it!

**Comment by druwadi** [comment]
This is perfect for our use case, great work! Looking forward to using this in a future release

**Comment by crenshaw-dev** [comment]
@komapa we'd be looking at 2.10 (Feb. 4), since we're already in the 2.9 feature freeze.

File: ui/src/app/applications/components/utils.tsx [hunk 3]
```diff
@@ -406,23 +406,44 @@ function getResourceActionsMenuItems(resource: ResourceTreeNode, metadata: model
...
+                                ),
+                                {
+                                    submit: async (vals, _, close) => {
+                                        try {
+                                            const resourceActionParameters = action.hasParameters ? [{name: action.name, value: vals.inputParameter}] : [];
+                                            await services.applications.runResourceAction(metadata.name, metadata.namespace, resource, action.name, resourceActionParameters);
+                                            close();
+                                        } catch (e) {
+                                            apis.notifications.show({
+                                                content: <ErrorNotification title='Unable to delete application' e={e} />,
```
Comment (crenshaw-dev) [inline ui/src/app/applications/components/utils.tsx:435]:
```suggestion
                                                content: <ErrorNotification title='Unable to run action' e={e} />,
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Looks like some of the styles/html still assumes a scale-specific action rather than a generic action. @irinam0992 are you still working on those changes?

**Pull request closed without merging**